### ISSUE TITLE
Cached content can not be fetched again?

### ISSUE THREAD

**Issue by DesignMonkey** [contentful-labs/cf-graphql#26]
Hi

I can see in the http-client.js's get method, that we test for cached content. But what if I make a change to my page inside Contentful? I can't refetch content unless i restart my server. Is there a setting i have missed, or is there a method to cleare cache? :) 

/Peter

**Comment by jelz** [comment]
Cache, by default, is created for each GraphQL request from the scratch. The reason is that sometimes a single GraphQL query can result in multiple calls to the same endpoint. In situations like that we want to reuse already issued HTTP calls.

Do you experience any problems or ask out of curiosity? If the former - please provide as much as possible (code of your server, Contentful content model, GraphQL queries you try) for me to reproduce the issue.

**Comment by DesignMonkey** [comment]
Well maybe it is not in this package I should find the problem then :) 

I'm using Next.js with React Apollo. My server.js code looks like this

`
const next = require('next')
const path = require('path');
const cfGraphql = require('cf-graphql');
const express = require('express');
const cors = require('cors');
const graphqlHTTP = require('express-graphql');

const port = process.env.PORT || 4000;
const spaceId = process.env.SPACE_ID;
const cdaToken = process.env.CDA_TOKEN;
const cmaToken = process.env.CMA_TOKEN;


const dev = process.env.NODE_ENV !== 'production'
const app = next({ dev })
const handle = app.getRequestHandler()


if (spaceId && cdaToken && cmaToken) {
  console.log('Space ID, CDA token and CMA token provided');
  console.log(`Fetching space (${spaceId}) content types to create a space graph`);
  useProvidedSpace();
} else {
  console.log('No env vars')
}

// this function implements a flow you could use in your application:
// 1. fetch content types
// 2. prepare a space graph
// 3. create a schema out of the space graph
// 4. run a server
function useProvidedSpace () {
  const client = cfGraphql.createClient({spaceId, cdaToken, cmaToken});

  client.getContentTypes()
    .then(cfGraphql.prepareSpaceGraph)
    .then(spaceGraph => {
      const names = spaceGraph.map(ct => ct.names.type).join(', ');
      console.log(`Contentful content types prepared: ${names}`);
      return spaceGraph;
    })
    .then(cfGraphql.createSchema)
    .then(schema => startServer(client, schema))
    .catch(fail);
}


function startServer (client, schema) {
  app.prepare().then(() => {
    const server = express()

    server.use('/graphql', graphqlHTTP({
      context: {entryLoader: client.createEntryLoader()},
      graphiql: false,
      schema,
    }));


    server.get('/episode/:urlname', (req, res) => {
      const actualPage = '/episode';
      const queryParams = { urlname: req.params.urlname }
      app.render(req, res, actualPage, queryParams)
    })

    server.get('*', (req, res) => {
      handle(req, res)
    })

    server.listen(3000)
  })
}

function fail (err) {
  console.log(err);
  process.exit(1);
}
`

But yea I'm experiencing this constantly... can't make the content to update after the first fetch.

Does it make any sense? :)

**Comment by jelz** [comment]
Quick check: do you modify your entries or content types?

**Comment by DesignMonkey** [comment]
I'm only fetching data, if that's answer your question?!

**Comment by jelz** [comment]
I'm asking about modification done in the Contentful web app (https://app.contentful.com/) - do you modify data in the "Content model" or "Content" tab?

![zrzut ekranu z 2017-07-24 22-53-01](https://user-images.githubusercontent.com/1936596/28544292-f6f472e4-70c2-11e7-8373-e77debd7ec4f.png)

**Comment by DesignMonkey** [comment]
arh.. now I understand. Sorry. I modify the entries. In the content tab.

And here is my code for the fetching:

```
import React, { Component } from 'react';
import { gql, graphql } from 'react-apollo';
import ReactMarkdown from 'react-markdown';

const EpisodeInfo = ({ data }) => {
  const { episodes } = data;

  if(data.loading) {
    return (
      <div>
        Loading...
      </div>
    )
  }
  if(!episodes || !episodes.length) {
    return (
      <div>
        Can't find urlname
      </div>
    )
  }
  const episode = episodes[0];
  return (
    <div>
      <h1>{episode.title}</h1>
      <ReactMarkdown source={episode.body} />
    </div>
  )
}

const episode = gql`
  query episode($query: String!) {
    episodes(q: $query) {
      uploadDate
      title
      urlname
      episode
      episodenum
      duration
      teaser
      soundcloud
      body
    }
  }
`

export default graphql(episode, {
  options: ({ urlname }) => {
    return {
      variables: {
        query: `fields.urlname=${urlname}`,
      }
    }
  },
  props: ({ data }) => {
    return {
      data,
    }
  }
})(EpisodeInfo)

```

**Comment by jelz** [comment]
Do you publish your entries after you modify them? `cf-graphql` uses Content Delivery API (https://www.contentful.com/developers/docs/references/content-delivery-api/) that serves only published changes. Please note that there is a small delay between publication and data being available (usually couple of seconds needed to distribute content to edge locations worldwide).

**Comment by DesignMonkey** [comment]
Yes I'm publishing, and i even after a nights sleep I couldn't make it to fetch the new data.

**Comment by DesignMonkey** [comment]
If I restart the server, it fetches the correct data again, but then that would be the cached version.

**Comment by jelz** [comment]
Hmmm, what happens if you replace:

```js
server.use('/graphql', graphqlHTTP({
  context: {entryLoader: client.createEntryLoader()},
  graphiql: false,
  schema,
}));
```

with:

```js
server.use('/graphql', graphqlHTTP(() => ({
  context: {entryLoader: client.createEntryLoader()},
  graphiql: false,
  schema,
})));
```

**Comment by DesignMonkey** [comment]
That did the trick. But why ? :) 

And thank you so much for your help :)

**Comment by jelz** [comment]
I took the liberty of reopening this issue. I'll extend the documentation and then close the ticket.

Quick explanation: if you pass an object to `graphqlHTTP` then all requests share exactly the same instance of `entryLoader` what causes too aggressive caching. If you pass a function to `graphqlHTTP` then a new `entryLoader` is created for each request what makes caching optimal (HTTP requests to the CDA are reused in a scope of the same GraphQL request).

As I said - I'll extend the docs to make it more clear. Maybe there's something that could be done on the code level to prevent situations like this.

Please note that the `expressGraphqlExtension` helper (https://github.com/contentful-labs/cf-graphql#helpers) takes care of forming an extension function in the right shape.

**Comment by DesignMonkey** [comment]
Sure thing. And it makes sense. Wouldn't have figured that one out myself :) 

I will try the expressGraphqlExtension helper. Once again. Thanks :)

**Comment by KevCJones** [comment]
Thanks for this, i have thoughts i want to share about this... i didn't think of just using a fresh instance, which was smarter than my blunt approach for a quick win. 

From a performance point of view, caching is valuable, as you said, some requests are multiple steps, and redoing those for data we know does not update much, might be infrequently changed. I want the best of both worlds i can get it.

Blunt solution of mine - I made a local copy and removed 3 lines from http-client.js

```
const cached = cache[url];
  // if (cached) {
  //   return cached;
  // }
```

## My perceived requirements

Always cache - Some data rarely changes and the cached responses are the same every time
Never cache -Some data is always changing (in my solution - i admit i'm probably leveraging more from cf than i it was meant for - later on we plan to integrate something like Social Engine and remove comments + likes from CF 👍 )
Conditional cache - Some data changes for reasonable reasons - we have articles, editors come in and publish updates, caching these would be nice for performance when they search with criteria already searched for, the data is returned much quicker.

1. Exposing cache object

The cache object looked like it could be fairly easily passed down via config and left to me to control. At the very least this would let me wipe out all/some cache when i want to without having to restart the server. 

2. Exposing a hook or callback that validates when a response is being cached. 

Ideally with a couple of small enhancements, we could expose when a cache  url and response are being stored so that some rudimentary but bespoke conditional checking can be done. When i get to implementing the idea i'd look at another config option for a function OR enhance the cache object to include some defined functions like `cache(url, response)`... not sure until i try it. 

With a hook like this, i could apply some simple regexp the url and applying some whitelist, blacklist stuff..

3. CF Webhooks can become useful

With exposed cache object on my side of the fence i can also make good use of the webhooks contentful provide so that in the case of things which infrequently change i can wipe out certain cached responses if for example, a collection like Articles is updated.. the webhook will trigger to an endpoint i define on my server and bam all cached queries `/entries?content_type=article...` are removed

If this is something you'd welcome input on, or attempted implementations i'll happily give it a go. It might be you have a better plan to achieve the same end?

**Comment by serkanyersen** [comment]
Agree with @KevCJones here. It would be really useful if we had more control over the cache. Regardless of the situation, you might always need the flush the cache during runtime.

Restarting the server is not an option during production.

@jelz I'm using the `expressGraphqlExtension` function but I still have this problem, responses are very aggressively cached. 

Thanks

**Comment by KevCJones** [comment]
If nothing else its useful for what i'm doing, if its deemed worth a merge, great. If not perhaps someone else can use it while the devs work on their own take.

**Comment by jelz** [comment]
@serkanyersen can you explain in more detail how your responses are "aggressively cached"? The idea is that all the responses are cached once and always reused through the course of a single request. Each GraphQL request gets its fresh data fetcher without anything cached.

@KevCJones I have to go through your solution to understand it better. If I understand correctly the idea is to cache infrequently changing entities. While I get the idea I'm not sure if it should be a part of this library. I'd prefer the HTTP client to have a well defined API so we can swap implementations. But I'll take a look at your solution next week.

**Comment by serkanyersen** [comment]
@jelz 

> Each GraphQL request gets its fresh data fetcher without anything cached.

it doesn't for me. I have to restart the server to get the fresh data. Unfortunately. 

It might be because how I setup the system. I had to make some changes in the code to accommodate routes being created the asynchronously, because I have a `*` route on my application.

Here is what my code looks like.

```javascript
const cfGraphql = require('cf-graphql');
const graphqlHTTP = require('express-graphql');
const config = require('./config');
const { logger } = require('./logger');

function setupGraphQL (app) {
  let ready;
  const readyPromise = new Promise(r => ready = r);

  const client = cfGraphql.createClient({
    spaceId: config.credentials.contentful.space_id,
    cdaToken: config.credentials.contentful.cda_token,
    cmaToken: config.credentials.contentful.cma_token,
  });

  // create routes synchronously, but wait till schema resolves  
  app.get('/graphiql', (_, res) => {
    // runs everytime /graphiql is hit.
    return readyPromise.then((opts) => {
      res.set(opts.ui.headers).status(opts.ui.statusCode).end(opts.ui.body);
      return opts;
    });
  });

  // graphqlHTTP accepts promise as asynchronous data as options.
  app.use('/graphql', graphqlHTTP(new Promise(resolve => {
    // runs once when setupGraphQL is executed. graphqlHTTP handles the route.
    readyPromise.then((opts) => {
      // get extensions config and set graphqlHTTP
      resolve(opts.ext());
    });
  })));

  client
    .getContentTypes()
    .then(cfGraphql.prepareSpaceGraph)
    .then((spaceGraph) => {
      const names = spaceGraph.map(ct => ct.names.type).join(', ');
      logger.log(`Contentful content types prepared: ${names}`);
      return spaceGraph;
    })
    .then(cfGraphql.createSchema)
    .then((schema) => {
      // activate the routes.
      ready({
        schema,
        ui: cfGraphql.helpers.graphiql({ title: 'contentful data' }),
        ext: cfGraphql.helpers.expressGraphqlExtension(client, schema, {
          version: true,
          timeline: config.isDev,
          detailedErrors: config.isDev,
        }),
      });

      logger.log('Running a GraphQL server!');
      logger.log(`You can access GraphiQL at ${config.domain}/graphiql/`);
      logger.log(`You can use the GraphQL endpoint at ${config.domain}/graphql/`);
    })
    .catch((err) => {
      logger.error(err);
    });
}

module.exports = {
  setupGraphQL,
};
```

**Comment by KevCJones** [comment]
@jelz

Thanks for getting back to us, appreciate it. I'll try to tackle each point made in turn and justify what i can.

> The idea is that all the responses are cached once and always reused through the course of a single request.

My experience was much like @serkanyersen.

Perhaps i was operating under a misunderstanding, but like above i found that my results were not reflecting the updates inside CF and when i looked into the codebase i saw [thought] the 'cache' code inside your http-client.js was the culprit. 

I see [above](https://github.com/contentful-labs/cf-graphql/issues/26#issuecomment-317556298) you made the recommendation to instantiate the client every call though i did setup my code much like it appeared in [https://github.com/contentful-labs/cf-graphql/blob/master/demo/server.js#L63](https://github.com/contentful-labs/cf-graphql/blob/master/demo/server.js#L63). 

Done like that, the cache object http-client maintains is persistent and thus we are where we are. 

>cached once and always reused through the course of a single request

Interesting, that would explain your cache object and where i stepped from the path, however, could you give me an example of a query using the apollo queries we can do right now where that cache would be used? I can only see by logging the key's that the full query url is keyed against the response, and since thats a round trip to the server and back to the client, i'm not seeing from my usage so far, where anything could be cached and reused during a single request of one query to `/graphql`?  I admit here, i'm relatively new to GraphQL, but just a high level use-case would put me on the trail, it could be something of use to me.

Also another bit of code that threw me was where [the codes initialisation of the 'cache' is something it checked config params at the http-client.js](https://github.com/contentful-labs/cf-graphql/blob/master/src/http-client.js#L17) level, suggesting it is something that could of been generated somewhere else and passed down into it. However [where you createClient](https://github.com/contentful-labs/cf-graphql/blob/48f0312b1b7157a67af7f54a7a855efea15b21e5/src/client.js) you choose not to offer up that option. 

> While I get the idea I'm not sure if it should be a part of this library.

Initially i wanted to build my own cache and not need to modify the library, but... having seen that the queries are posted to the /graphql are processed within cf-graphql's middleware solution. OR intercept responses easily to maintain my own cache (given a certain amount of nodeJS novice-ness on my part) and so the option without my own tinkered solution is to either completely lose any cache optimisations [by re-instantiating the graphHTTP object each time OR implement a way to selectively cache and keep the cache on my side of the library (if i chose to).

Perhaps you could offer up an implementation that would let me simply maintain a cache inside my own server.js? I'd be happy with that. I actually think a bare minimum option could be to simply offer us a config parameter to pass in the cache object... without the `shouldCache` nice to have. With a cache object on my side of the fence, i can simply leverage the webhooks to choose how and when to uncache items. 

### Did it actually make a difference?

TL;DR; Yes - without the cache time to first byte were 50-160%  worse off from very limited testing.

Some latency checks (very unscientific - i'd have to give you internet speeds and equipment specs and servers etc etc.... but this is just for comparative gains). I just run the request a few times and take the average time to first byte from chromes dev tools.

Query #1 for a list of disease models, very small call, rarely changes - 500bytes of payload data. 
- With caching 55ms TTFB
- Without caching 85ms TTFB 

So its a 50% increase in smaller queries.

Query #2 a document (article) with 6000bytes of data to pull down, los of text and markdown.
- with cache on 55ms
- without cache 145ms

A 163% increase.

The added gains here are seen as well because concurrent users benefit from the caching too. 

I don't think my solution is a bad one per say, i do think there is some middle ground where we can still have a win win though.

**Comment by jelz** [comment]
I think we're mixing two things.

This library has no intention to do any high-level caching since we don't want to assume anything about usage, cache invalidation, frequency of entity changes and so on. Ideally there should be a mechanism to provide a customized fetching logic that may include some kind of global caching.

The `cache` used in the current HTTP client is a low-level cache that:

1. lives only throughout the lifetime of a single GraphQL request
2. eliminates obviously repeated calls, see this dummy example:

```graphql
{
  c1: category(id: "6XL7nwqRZ6yEw0cUe4y0y6") {
    title
  }
  
  c2: category(id: "6XL7nwqRZ6yEw0cUe4y0y6") {
    title
  }
}
```

The `cache` prevents the second call. That's about it when it comes to "caching" implemented in the library so far.

**Comment by jelz** [comment]
@serkanyersen can you try passing a function to express GraphQL middleware? I think it could help. Somtething like that:

`  app.use('/graphql', graphqlHTTP(() => new Promise(resolve => {`

**Comment by jelz** [comment]
@serkanyersen I investigated further and it seems you're using `express-graphql` incorrectly. Here's a fixed version of your code:

```js
function setupGraphQL (app, config) {
  let ready;
  const readyPromise = new Promise(r => ready = r);

  const client = cfGraphql.createClient(config);
  const ui = cfGraphql.helpers.graphiql({ title: 'contentful data' });

  // create routes synchronously, but wait till schema resolves
  app.get('/graphiql', (_, res) => readyPromise.then(() => {
    res.set(ui.headers).status(ui.statusCode).end(ui.body);
  }));

  // graphqlHTTP accepts promise as asynchronous data as options.
  app.use('/graphql', graphqlHTTP(() => readyPromise.then(schema => {
    const ext = cfGraphql.helpers.expressGraphqlExtension(client, schema, {
      version: true,
      timeline: true,
      detailedErrors: true
    });

    return ext();
  })));

  client
    .getContentTypes()
    .then(cfGraphql.prepareSpaceGraph)
    .then((spaceGraph) => {
      const names = spaceGraph.map(ct => ct.names.type).join(', ');
      console.log(`Contentful content types prepared: ${names}`);
      return spaceGraph;
    })
    .then(cfGraphql.createSchema)
    .then((schema) => {
      // activate the routes.
      ready(schema);
      console.log('Running');
    })
    .catch((err) => {
      console.error(err);
    });
}
```

**Comment by KevCJones** [comment]
@jelz thanks for your explanations above - i've seen where i left the path now and i'm back on it.

Manage my own cache solution, ignore the internal cache ;) a basic combination of memory-cache and webhooks to flush it out if someone publishes changes. 

Only down side of this approach for me is minor - your internal cache urls are clear CDA requests... but from a middleware approach like this, i'm having to bodyparse and then concatinate from the query + variables a key to cache the response against... its a bit fugly to see in the console while i'm testing... but it does the job. 

Thanks again for steering me in the right direction.

**Comment by jelz** [comment]
Detailed information about construction of `context` and per-request entry loader instances included in the README of v0.5.

High level cache is a separate feature and I want to get the basics right first.
